Clean up ReducedBeamCharacteristics with TypeMultiplier

n01r opened the pull request:
The `ReducedBeamCharacteristics` reduced diagnostics had a lot of code where `amrex::ReduceOps` were constructed from lists of types. This PR cleans it up with the `TypeMultiplier` utility that makes it more legible.

Credit goes to @AlexanderSinn's work in https://github.com/Hi-PACE/hipace/pull/1052 and https://github.com/AMReX-Codes/amrex/pull/3718.

Same changes as in: https://github.com/ECP-WarpX/WarpX/pull/4798

n01r committed:
Clean up RedBeamChar with TypeMultiplier

Make `ReducedBeamCharacteristics` more readable.

n01r committed:
Change length variable to constexpr size_t

The variables `num_red_ops_*` are used for the TypeMultiplier and also for
the length of the output vectors for the reduced diagnostics.
This commit changes their type from `const int` to `constexpr size_t` because
this is more appropriate for a vector length, and with constexpr we indicate
that the value is constant and can be evaluated at compile time already.

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -136,23 +108,25 @@ namespace impactx::diagnostics
+         * px_min, py_min, pt_min
+         */
+        amrex::constexpr_for<0, num_red_ops_1_min> ([&](auto i) {
+            values_per_rank_min[i] = amrex::get<i + num_red_ops_1_sum>(r1);
```
I see this in HiPACE++:
```suggestion
            values_per_rank_min[i.value] = amrex::get<i.value + num_red_ops_1_sum>(r1);
```

AlexanderSinn replied:
I didn’t know that `std::integral_constant` had a conversion operator to the integer. Without `.value` should work as well. Not sure if this is related to the failing MSVC CI test or not.

ax3l replied:
Copying error for records:
```
     8>D:\a\impactx\impactx\src\particles\diagnostics\ReducedBeamCharacteristics.cpp(118,45): error C2672: 'amrex::get': no matching overloaded function found [D:\a\impactx\impactx\build\lib.vcxproj]
         D:\a\impactx\impactx\build\_deps\fetchedamrex-src\Src\Base\AMReX_Tuple.H(196,1):
         could be 'GpuTupleElement<I,amrex::GpuTuple<Ts...>>::type &&amrex::get(amrex::GpuTuple<Ts...> &&) noexcept'
         	D:\a\impactx\impactx\src\particles\diagnostics\ReducedBeamCharacteristics.cpp(118,51):
         	expression did not evaluate to a constant
         		D:\a\impactx\impactx\src\particles\diagnostics\ReducedBeamCharacteristics.cpp(118,53):
         		failure was caused by a read of a variable outside its lifetime
         		D:\a\impactx\impactx\src\particles\diagnostics\ReducedBeamCharacteristics.cpp(118,53):
         		see usage of 'this'
```

column 51 refers to the values inside `get< ... >`

ax3l replied:
@n01r:
```suggestion
            constexpr int idx = i.value + num_red_ops_1_sum;
            values_per_rank_min[i] = amrex::get<idx>(r1);
```

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -101,25 +77,21 @@ namespace impactx::diagnostics
+         * px_mean, py_mean, pt_mean
+         */
+        amrex::constexpr_for<0, num_red_ops_1_sum> ([&](auto i) {
+            values_per_rank_1st[i] = amrex::get<i>(r1);
```
I see this in HiPACE++:
```suggestion
            values_per_rank_1st[i.value] = amrex::get<i.value>(r1);
```

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -136,23 +108,25 @@ namespace impactx::diagnostics
+         * px_max, py_max, pt_max
+         */
+        amrex::constexpr_for<0, num_red_ops_1_max> ([&](auto i) {
+            values_per_rank_max[i] = amrex::get<i + num_red_ops_1_sum + num_red_ops_1_min>(r1);
```
```suggestion
            values_per_rank_max[i.value] = amrex::get<i.value + num_red_ops_1_sum + num_red_ops_1_min>(r1);
```

n01r committed:
Add constexpr std::size_t indices for amrex::get

We had MSVC (only on Windows) compile errors and this commit tries to fix them.

n01r commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -136,23 +108,27 @@
+         * px_max, py_max, pt_max
+         */
+        amrex::constexpr_for<0, num_red_ops_1_max> ([&](auto i) {
+            constexpr std::size_t idx = i + num_red_ops_1_sum + num_red_ops_1_min;
```
This is a false positive due to non-traditional variable use that the analysis tool identified and we can ignore it, right @ax3l?

ax3l reviewed (approved):
Thank you! :)
